test/functional/p2p_orphan_handling.py

```diff
@@ -0,0 +1,416 @@
+        self.relay_transaction(peer1, parent1_witness_stripped)
+        assert_equal(parent_normal["txid"], parent1_witness_stripped.rehash())
+        assert parent1_witness_stripped.rehash() not in node.getrawmempool()
+
+        # Relay the child. It should not be accepted because it has missing inputs.
+        self.relay_transaction(peer2, child_invalid_witness["tx"])
+        assert child_invalid_witness["txid"] not in node.getrawmempool()
+
+        # The parent should be requested since the unstripped wtxid would differ. Delayed because
+        # it's by txid and this is not a preferred relay peer.
+        self.nodes[0].bumpmocktime(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        peer2.wait_for_getdata([int(parent_normal["tx"].rehash(), 16)])
+
+        # parent_normal can be relayed again even though parent1_witness_stripped was rejected
+        self.relay_transaction(peer1, parent_normal["tx"])
+        assert_equal(set(node.getrawmempool()), set([parent_normal["txid"], child_invalid_witness["txid"]]))
+
+    @cleanup
+    def test_orphan_multiple_parents(self):
+        node = self.nodes[0]
+        peer = node.add_p2p_connection(PeerTxRelayer())
+
+        self.log.info("Test orphan parent requests with a mixture of confirmed, in-mempool and missing parents")
+        # This UTXO confirmed a long time ago.
+        utxo_conf_old = self.wallet.send_self_transfer(from_node=node)["new_utxo"]
+        txid_conf_old = utxo_conf_old["txid"]
+        self.generate(self.wallet, 10)
+
+        # Create a fake reorg to trigger BlockDisconnected, which resets the rolling bloom filter.
+        # The alternative is to mine thousands of transactions to push it out of the filter.
+        last_block = node.getbestblockhash()
+        node.invalidateblock(last_block)
+        node.preciousblock(last_block)
+        node.syncwithvalidationinterfacequeue()
+
+        # This UTXO confirmed recently.
+        utxo_conf_recent = self.wallet.send_self_transfer(from_node=node)["new_utxo"]
+        self.generate(node, 1)
+
+        # This UTXO is unconfirmed and in the mempool.
+        assert_equal(len(node.getrawmempool()), 0)
+        mempool_tx = self.wallet.send_self_transfer(from_node=node)
+        utxo_unconf_mempool = mempool_tx["new_utxo"]
+
+        # This UTXO is unconfirmed and missing.
+        missing_tx = self.wallet.create_self_transfer()
+        utxo_unconf_missing = missing_tx["new_utxo"]
+        assert missing_tx["txid"] not in node.getrawmempool()
+
+        orphan = self.wallet.create_self_transfer_multi(utxos_to_spend=[utxo_conf_old,
+            utxo_conf_recent, utxo_unconf_mempool, utxo_unconf_missing])
+
+        self.relay_transaction(peer, orphan["tx"])
+        self.nodes[0].bumpmocktime(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        peer.sync_with_ping()
+        assert_equal(len(peer.last_message["getdata"].inv), 2)
+        peer.wait_for_parent_requests([int(txid_conf_old, 16), int(missing_tx["txid"], 16)])
```

Comment:
Requesting the already confirmed transaction is surprising to me. Is this expected to change in future work (if so a comment mentioning that would be useful) or is it intended behavior?

Reply:
> Requesting the already confirmed transaction is surprising to me.

Yep, was strange to me too, but `AlreadyHaveTx` is peerman's best metric for determining what the missing parents are (validation doesn't say which one(s) are missing). Also afaiu it's considered uncommon to (1) have an orphan (2) have a transaction that spends a very old coin, so I'm guessing this is quite rare.

> Is this expected to change in future work

I'm not aware of any plans to change the legacy way of requesting orphan parents. But with package relay we'd ask for the list of unconfirmed ancestors, so we wouldn't have this kind of false positive.